### PR TITLE
[sui-framework] Remove non-delegation based staking flows

### PR DESCRIPTION
**Pull request by tzakian**
## Description 

This unifies the staking flows for a validator so there is no distinction between self and delegated stake. This additionally cleans up a number of object fields that are no longer needed after this change.

This is a breaking change since it changes the object representation for `Validator` objects, as well as the validator staking/unstaking flows.

## Test Plan 

Updated existing framework tests
---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [X] user-visible impact
- [X] breaking change for a client SDKs
- [?] breaking change for FNs (FN binary must upgrade)
- [X] breaking change for validators or node operators (must upgrade binaries)
- [X] breaking change for on-chain data layout
- [X] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-framework/sources/governance/validator_set.move [hunk 1]
```diff
@@ -24,12 +23,9 @@ module sui::validator_set {
     friend sui::validator_set_tests;
 
     struct ValidatorSet has store {
-        /// Total amount of stake from all active validators (not including delegation),
+        /// Total amount of stake from all active validators (including delegation),
```
Comment (sblackshear) [inline crates/sui-framework/sources/governance/validator_set.move:26]:
```suggestion
        /// Total amount of stake from all active validators
```

Probably makes sense to drop this now that delegated stake is the only thing

**Review by sblackshear — COMMENTED** [review]
Nice!!

One note: this will probably break the `old_frenemies`/`new_frenemies` examples. Totally ok to delete one of them to reduce the amount of code you need to fix if it simplifies things.

File: crates/sui-framework/tests/validator_tests.move [hunk 2]
```diff
@@ -89,34 +88,36 @@ module sui::validator_tests {
...
-            let stake = test_scenario::take_from_sender<Stake>(scenario);
+            let delegation = test_scenario::take_from_sender<Delegation>(scenario);
+            let stake = test_scenario::take_from_sender_by_id<StakedSui>(scenario, *staking_pool::staked_sui_id(&delegation));
             let ctx = test_scenario::ctx(scenario);
-            validator::request_withdraw_stake(&mut validator, &mut stake, 5, 35, ctx);
-            test_scenario::return_to_sender(scenario, stake);
-            assert!(validator::stake_amount(&validator) == 10, 0);
+            validator::request_withdraw_delegation(&mut validator, delegation, stake, ctx);
+
+            assert!(validator::total_stake(&validator) == 10, 0);
```
Comment (emmazzz) [inline crates/sui-framework/tests/validator_tests.move:104]: Not for this PR but in the future we can start using the nice `test_utils::assert_eq` function :D

File: crates/sui-framework/tests/rewards_distribution_tests.move [hunk 3]
```diff
@@ -249,15 +249,15 @@ module sui::rewards_distribution_tests {
         // unslashed validators too.
         // Validator 4 should get 475 SUI of rewards without slashing. 20% is slashed so she gets
         // 380 SUI of rewards.
-        assert_validator_stake_amounts(validator_addrs(), vector[294, 508, 722, 780], scenario);
+        assert_validator_self_stake_amounts(validator_addrs(), vector[294, 508, 723, 780], scenario);
```
Comment (emmazzz) [inline crates/sui-framework/tests/rewards_distribution_tests.move:252]: I suppose this and the change below are due to rounding differences? And we are pretty sure why this change happened?

File: crates/sui-framework/tests/rewards_distribution_tests.move [hunk 4]
```diff
@@ -85,17 +79,23 @@ module sui::rewards_distribution_tests {
...
-        assert_validator_stake_amounts(validator_addrs(), vector[110, 220, 330, 440], scenario);
+        assert_validator_total_stake_amounts(validator_addrs(), vector[330, 330, 330, 440], scenario);
+        assert_validator_self_stake_amounts(validator_addrs(), vector[110, 220, 330, 440], scenario);
         undelegate(DELEGATOR_ADDR_1, 0, 0, scenario);
         delegate_to(DELEGATOR_ADDR_2, VALIDATOR_ADDR_1, 600, scenario);
         // 10 SUI rewards for each 110 SUI of stake
         advance_epoch_with_reward_amounts(0, 130, scenario);
-        assert!(total_sui_balance(DELEGATOR_ADDR_1, scenario) == 240, 0); // 40 SUI of rewards received
-        assert_validator_stake_amounts(validator_addrs(), vector[120, 240, 360, 480], scenario);
+        // 40 SUI of rewards received by the validator
```
Comment (emmazzz) [inline crates/sui-framework/tests/rewards_distribution_tests.move:93]: Should be "received by the delegator"?

File: crates/sui-framework/sources/governance/staking_pool.move [hunk 5]
```diff
@@ -424,9 +422,37 @@ module sui::staking_pool {
...
+            let elem_key = *option::borrow(&elem_opt);
+            let pending_delegation_entry = linked_table::borrow(&staking_pool.pending_delegations, elem_key);
+            sum = sum + pending_delegation_entry.sui_amount;
+            elem_opt = *linked_table::next(&staking_pool.pending_delegations, elem_key);
+        };
+        sum
+    }
+
+    /// Calculate the current the total withdrawal requests (in terms of principal) for the staking pool
+    public fun pending_principal_withdrawal_amounts(staking_pool: &StakingPool): u64 {
```
Comment (emmazzz) [inline crates/sui-framework/sources/governance/staking_pool.move:441]: This and the function above walk through the (potentially pretty big) table every time and do a lot of dynamic child loading. Maybe keep a rolling sum in the struct instead?

File: crates/sui-framework/sources/governance/validator.move [hunk 6]
```diff
@@ -389,7 +344,11 @@ module sui::validator {
             0
         );
         let stake_amount = balance::value(&stake);
-        stake::create(stake, sui_address, coin_locked_until_epoch, ctx);
+        let staking_pool = staking_pool::new(sui_address, tx_context::epoch(ctx) + 1, ctx);
+        // Add the validator's starting stake to the staking pool.
+        staking_pool::request_add_delegation(&mut staking_pool, stake, coin_locked_until_epoch, sui_address, ctx);
+        // We immediately process this delegation as they are at validator setup time and this is the validator staking with itself.
+        staking_pool::process_pending_delegations(&mut staking_pool, ctx);
```
Comment (emmazzz) [inline crates/sui-framework/sources/governance/validator.move:351]: Nice!

File: crates/sui-framework/tests/validator_tests.move (hunk 2 above)
Comment (tzakian) [reply crates/sui-framework/tests/validator_tests.move:104]: Yes! That would be amazing :) Happy to make a pass through at some point and convert them over :)

File: crates/sui-framework/tests/rewards_distribution_tests.move (hunk 3 above)
Comment (tzakian) [reply crates/sui-framework/tests/rewards_distribution_tests.move:252]: I believe these are indeed due to rounding differences, and in particular because we no longer scale the validators commission by the validator's self stake. But I want to dive into this change a bit more and be confident in _exactly_ why they before I would call this ready :)

File: crates/sui-framework/sources/governance/staking_pool.move (hunk 5 above)
Comment (tzakian) [reply crates/sui-framework/sources/governance/staking_pool.move:441]: Good question! I believe (and will need to double-check) that these functions are only used in tests, so not sure if we should add a field to the structs to keep the rolling sum. If you think these will be useful to have in general as non-test helpers, happy to make this change :)

File: crates/sui-framework/tests/rewards_distribution_tests.move (hunk 3 above)
Comment (tzakian) [reply crates/sui-framework/tests/rewards_distribution_tests.move:252]: Dug into this a bit more this morning and have some interesting conclusions:
1. This is _not_ because of rounding issues during rewards distribution -- if I look at the numbers getting distributed to each validator during rewards distribution before and after the changes in this PR they are the exact same same for this test. 
2. What is causing this change is the logic in the test function `calculate_rewards`:
    a. We are calculating the correct SUI amount for the deposited pool tokens (`79` in this case).
    b. However, the staked amount is `80`. So the logic in that function calculates the reward amount as `80`.

Now, I'm not sure if this is correct or not. I was simply copying the logic around `withdraw_rewards_and_burn_pool_tokens`. So I'll try to understand that a bit more this morning, but I may want to chat with you about it a bit at some point :)

File: crates/sui-framework/sources/governance/staking_pool.move (hunk 5 above)
Comment (emmazzz) [reply crates/sui-framework/sources/governance/staking_pool.move:441]: Let's keep it simple and make it test only then :)

Comment (tzakian) [reply crates/sui-framework/sources/governance/staking_pool.move:437]: Actually had a look and it appears as though this function is used by `frenemies` so I think we might need to add this field in unless we want to depricate that functionality over there 😢

Comment (tzakian) [reply crates/sui-framework/sources/governance/staking_pool.move:306]: Seems like frenemies is getting a soft deprecation, so I vote we keep this as-is for now, and we can remove later if that sounds good?

**Review by Jordan-Mysten — APPROVED** [review]
TS changes LGTM

File: crates/sui-framework/sources/governance/validator_set.move [hunk 7]
```diff
@@ -364,23 +322,19 @@ module sui::validator_set {
...
     public fun validator_total_stake_amount(self: &ValidatorSet, validator_address: address): u64 {
         let validator = get_validator_ref(&self.active_validators, validator_address);
         validator::total_stake_amount(validator)
     }
 
-    public fun validator_stake_amount(self: &ValidatorSet, validator_address: address): u64 {
-        let validator = get_validator_ref(&self.active_validators, validator_address);
-        validator::stake_amount(validator)
-    }
+    //public fun validator_stake_amount(self: &ValidatorSet, validator_address: address): u64 {
```
Comment (emmazzz) [inline crates/sui-framework/sources/governance/validator_set.move:334]: Remove this?

**Review by emmazzz — APPROVED** [review]
Look beautiful to me!! Thank you!
